packages/studio-base/src/panels/ThreeDeeRender/renderables/CameraStateSettings.ts

```diff
@@ -4,113 +4,250 @@
 
 import { t } from "i18next";
 import { cloneDeep, set } from "lodash";
+import * as THREE from "three";
+import { OrbitControls } from "three/examples/jsm/controls/OrbitControls";
 
 import { SettingsTreeAction } from "@foxglove/studio";
+import {
+  AnyFrameId,
+  CoordinateFrame,
+  Pose,
+  UserFrameId,
+  makePose,
+} from "@foxglove/studio-base/panels/ThreeDeeRender/transforms";
 
-import { Renderer } from "../Renderer";
+import { FollowMode, Renderer } from "../Renderer";
 import { SceneExtension } from "../SceneExtension";
 import { SettingsTreeEntry } from "../SettingsManager";
-import { DEFAULT_CAMERA_STATE } from "../camera";
-import { PRECISION_DEGREES, PRECISION_DISTANCE } from "../settings";
+import { CameraState, DEFAULT_CAMERA_STATE } from "../camera";
+import { PRECISION_DEGREES, PRECISION_DISTANCE, SelectEntry } from "../settings";
 
+const UNIT_X = new THREE.Vector3(1, 0, 0);
+const UNIT_Z = new THREE.Vector3(0, 0, 1);
+const PI_2 = Math.PI / 2;
+
+// used for holding unfollowPoseSnapshot in render frame every new frame
+const snapshotInRenderFrame = makePose();
+
+const tempVec3 = new THREE.Vector3();
+
+const tempSpherical = new THREE.Spherical();
+const tempEuler = new THREE.Euler();
+const FOLLOW_TF_PATH = ["general", "followTf"];
 export class CameraStateSettings extends SceneExtension {
-  public constructor(renderer: Renderer) {
+  // The pose of the render frame in the fixed frame when following was disabled
+
+  private unfollowSnapshotFrameIds:
```

Comment:
```suggestion
  /** The pose of the render frame in the fixed frame when following was disabled */
  private unfollowSnapshotFrameIds:
```

maybe add more on what it's used for?